SyntaxGym benchmark with multi-suite capability

jimn2 opened the pull request:
I 1. checked out the old branch, 2. saved the files, 3. checked out main, 4. updated main locally, 5.  created this new branch (jim/syntaxgym-2), 6. copied all files that were new or differed from those in main, 7. pushed and opened this PR.

In my last update I added capability to evaluate many suite files instead of just one by specifying a list of suite paths and it's passing test_integration.py for both single and multi-suite cases.

I'm going to switch the suite path list to a dict that contains the suite name and path.

hans commented:
Fixed numerical mismatch with #95 . This test will pass after that is merged and we can rebase.

jimn2 commented:
Fixed a bug in the last committed version where the benchmark was attempting to return a floating point value rather than a Score assembly which was causing an error.  Also added the capability to load test_suites directly from GitHub repos (or other URLs) so there's no longer any need to store test_suites locally.

mschrimpf commented on brainscore_language/benchmarks/syntaxgym/test_suites.py:
```diff
@@ -0,0 +1,108 @@
+suite33 = 'https://raw.githubusercontent.com/cpllab/syntactic-generalization/nextflow/test_suites/json/subordination_pp-pp.json'
+suite34 = 'https://raw.githubusercontent.com/cpllab/syntactic-generalization/nextflow/test_suites/json/subordination_src-src.json'
+
+test_suite_dict = {name1: suite1,
```
I think it would be more readable to just write it out directly, i.e.
```
test_suite_dict = {
    'center_embed': 'https://raw.githubusercontent.com/cpllab/syntactic-generalization/nextflow/test_suites/json/center_embed.json',
    'center_embed_mod': 'https://raw.githubusercontent.com/cpllab/syntactic-generalization/nextflow/test_suites/json/center_embed_mod.json',
    ...
```
etc.

Looking at this, couldn't you even just define the names, i.e. 
```
test_suite_names = ['center_embed', 'center_embed_mod', ...]
```
etc and then build the urls dynamically, i.e.
```
test_suite_dict = {name: f'https://raw.githubusercontent.com/cpllab/syntactic-generalization/nextflow/test_suites/json/{name}.json'}
```

mschrimpf commented on brainscore_language/benchmarks/syntaxgym/test_suites.py:
```diff
@@ -0,0 +1,108 @@
+suite31 = 'https://raw.githubusercontent.com/cpllab/syntactic-generalization/nextflow/test_suites/json/subordination.json'
+suite32 = 'https://raw.githubusercontent.com/cpllab/syntactic-generalization/nextflow/test_suites/json/subordination_orc-orc.json'
+suite33 = 'https://raw.githubusercontent.com/cpllab/syntactic-generalization/nextflow/test_suites/json/subordination_pp-pp.json'
+suite34 = 'https://raw.githubusercontent.com/cpllab/syntactic-generalization/nextflow/test_suites/json/subordination_src-src.json'
```
Can we index these by commit so that we always reference the same files, even if commits are made?
E.g. (this is a slightly different url, but there should be a way to get the commit hash in there; on github by simply pressing `y`):
```
https://github.com/cpllab/syntactic-generalization/raw/43f1f86b746254406dafc89bd17b9d4976607ef6/test_suites/json/center_embed.json
```

jimn2 commented on brainscore_language/benchmarks/syntaxgym/__init__.py:
```diff
@@ -0,0 +1,133 @@
+    return SyntaxGymTSE(suite_paths)
+
+
+class SyntaxGymTSE(BenchmarkBase):
```
Yes.  I think there are three relevant papers right?  The first was Jen Hu, the last was Ethan Wilcox, and I can't remember what the second one was, but I'll ask Ethan.

mschrimpf replied:
At least Hu & Gauthier; @hans should say if we need to cite other papers for this set of engineering benchmarks (Wilcox is behavioral one afaik, i.e. should be separate)

jimn2 replied:
I talked with Ethan about this and referenced what he called the main SyntaxGym paper.  (2020-Gauthier, Hu, Wilcox, etal)

jimn2 commented on brainscore_language/benchmarks/syntaxgym/__init__.py:
```diff
@@ -0,0 +1,133 @@
+        Compute region-level surprisal totals for the given subject.
+        """
+        suite_regions = list(self.suite.iter_regions())
+        candidate.start_behavioral_task(task=ArtificialSubject.Task.reading_times)
```
fixed.

jimn2 commented on brainscore_language/benchmarks/syntaxgym/__init__.py:
```diff
@@ -0,0 +1,133 @@
+            bibtex=None)
+
+        self.metric = load_metric('accuracy')
+        # TODO support non-path ref
```
Martin, did you add this todo?

mschrimpf replied:
no

jimn2 commented on brainscore_language/benchmarks/syntaxgym/__init__.py:
```diff
@@ -0,0 +1,133 @@
+        # TODO support non-path ref
+        self.suite = self._load_suite(suite_ref)
+
+    def _load_suite(self, suite_ref):
```
Yes.

jimn2 commented on brainscore_language/benchmarks/syntaxgym/test.py:
```diff
@@ -0,0 +1,1193 @@
+# May include subsets of actual items for speed.
+#
+# See notebook in https://colab.research.google.com/drive/1qziyPcu65jffizSPi-ZGHKR0x7BaHFMS#scrollTo=RgtnScy6LLKi .
+REFERENCE_DISTILGPT2_REGION_TOTALS = {
```
the keys in those reference region totals are tuples which include ints.  json keys have to be strings.  i'll look into other file formats.

kvfairchild commented on brainscore_language/benchmarks/syntaxgym/test.py:
```diff
@@ -0,0 +1,1193 @@
+
+@pytest.fixture(scope="session")
+def distilgpt2():
+  return HuggingfaceSubject(model_id="distilgpt2", region_layer_mapping={})
```
It's not ideal but it shouldn't be an issue per se. The other plugin's dependencies will be installed into the `conda` env during testing, so it's no longer a clean isolated environment just to test your plugin. This is philosophically maybe undesirable but the only headache I can anticipate would be your tests failing due to some future conflict that has nothing to do with your plugin.

mschrimpf replied:
Ok, the initial way of `return HuggingfaceSubject(model_id="distilgpt2", region_layer_mapping={})` might then actually be preferable

hans commented on brainscore_language/benchmarks/syntaxgym/__init__.py:
```diff
@@ -62,10 +62,23 @@ def __init__(self, suite_ref_list):
+        from pprint import pprint
+        pprint(sub_scores)
+
+        sub_scores = Score.merge(*sub_scores)
```
`AttributeError: 'numpy.ndarray' object has no attribute 'rename'`

This doesn't work at the moment and I don't understand why. Can someone with `xarray` experience help out here?
All the `sub_scores` are `DataArray` values.

There's a corresponding test `brainscore_language/benchmarks/syntaxgym/test.py:test_score_aggregate` you can use to test this yourself. Here's the current test output:

```
=========================================================================================== FAILURES ===========================================================================================
_____________________________________________________________________________________ test_score_aggregate _____________________________________________________________________________________
                                                                                                                                                                                                
distilgpt2 = <brainscore_language.model_helpers.huggingface.HuggingfaceSubject object at 0x7ff0b0cec0d0>                                                                                        
                                                                                                                                                                                                
    def test_score_aggregate(distilgpt2):                                                                                                                                                       
        suites = ["cleft", "cleft_modifier"]                                                                                                                                                    
        benchmark = SyntaxGymTSE(suites)                                                                                                                                                        
                                                                                                                                                                                                
>       score = benchmark(distilgpt2)                                                                                                                                                           
                                                                                                                                                                                                
brainscore_language/benchmarks/syntaxgym/test.py:1155:                                                                                                                                          
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
brainscore_language/benchmarks/syntaxgym/__init__.py:77: in __call__                                                                                                                            
    sub_scores = Score.merge(*sub_scores)
../../miniconda3/envs/brainscore/lib/python3.10/site-packages/brainscore_core/metrics/__init__.py:97: in merge
    raise e
../../miniconda3/envs/brainscore/lib/python3.10/site-packages/brainscore_core/metrics/__init__.py:90: in merge
    attr_values = Score.merge(*attr_values, ignore_exceptions=True)
../../miniconda3/envs/brainscore/lib/python3.10/site-packages/brainscore_core/metrics/__init__.py:85: in merge
    result = merge_data_arrays(scores)
../../miniconda3/envs/brainscore/lib/python3.10/site-packages/brainio/assemblies.py:322: in merge_data_arrays
    merged = xr.merge((data_array.rename('z') for data_array in data_arrays))['z'].rename(None)
../../miniconda3/envs/brainscore/lib/python3.10/site-packages/xarray/core/merge.py:895: in merge
    for obj in objects:
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

.0 = <tuple_iterator object at 0x7ff0aef8c7c0>

>   merged = xr.merge((data_array.rename('z') for data_array in data_arrays))['z'].rename(None)
E   AttributeError: 'numpy.ndarray' object has no attribute 'rename'

../../miniconda3/envs/brainscore/lib/python3.10/site-packages/brainio/assemblies.py:322: AttributeError
------------------------------------------------------------------------------------- Captured stdout call -------------------------------------------------------------------------------------
[<xarray.Score (sub_benchmark: 1)>
array([0.25])
Coordinates:
  * sub_benchmark  (sub_benchmark) <U5 'cleft'
Attributes:
    raw:      [False False  True  True False False False False False False Fa...,
 <xarray.Score (sub_benchmark: 1)>
array([0.35])
Coordinates:
  * sub_benchmark  (sub_benchmark) <U14 'cleft_modifier'
Attributes:
    raw:      [False  True False  True False  True False False False  True  T...]
```

mschrimpf replied:
just pushed a fix, see if that works. You will need to also update `core` for it to work.

mschrimpf commented on brainscore_language/benchmarks/syntaxgym/__init__.py:
```diff
@@ -0,0 +1,193 @@
+# See the SyntaxGym website for information about structuring test_suites:
+# https://cpllab.github.io/syntaxgym-core/architecture.html
+#
+# @inproceedings{gauthier-etal-2020-syntaxgym,
+#     title = "{S}yntax{G}ym: An Online Platform for Targeted Evaluation of Language Models",
+#     author = "Gauthier, Jon, etal
+#     booktitle = "Proceedings of the 58th Annual Meeting of the Association for Computational Linguistics: System Demonstrations",
+#     year = "2020",
+#     publisher = "Association for Computational Linguistics",
+#     url = "https://aclanthology.org/2020.acl-demos.10",
+#     abstract = "Targeted syntactic evaluations have yielded insights into the generalizations learned by neural network language
+#     models. However, this line of research requires an uncommon confluence of skills: both the theoretical knowledge needed to
+#     design controlled psycholinguistic experiments, and the technical proficiency needed to train and deploy large-scale language
+#     models. We present SyntaxGym, an online platform designed to make targeted evaluations accessible to both experts in NLP
+#     and linguistics, reproducible across computing environments, and standardized following the norms of psycholinguistic
+#     experimental design. This paper releases two tools of independent value for the computational linguistics
+#     community: 1. A website, syntaxgym.org, which centralizes the process of targeted syntactic evaluation and provides
+#     easy tools for analysis and visualization; 2. Two command-line tools, {`}syntaxgym{`} and {`}lm-zoo{`}, which allow
+#     any user to reproduce targeted syntactic evaluations and general language model inference on their own machine."}
```
recommend putting this in a `BIBTEX` variable

mschrimpf commented on brainscore_language/benchmarks/syntaxgym/__init__.py:
```diff
@@ -0,0 +1,193 @@
+    # Create a benchmark with matching suites.
+    matching_suites = [suite_name for suite_name in test_suite_dict.keys()
+                       if suite_name.startswith(tuple(suite_tags))]
+    benchmark_registry[f"syntaxgym-2020/{circuit}"] = lambda: SyntaxGymTSE(matching_suites)
```
@kvfairchild as far as I can tell, this will not play straight-forwardly with the plugin identifier matching via code parsing. How should we proceed here?

mschrimpf reviewed (commented):
These are the identifiers I get in the `benchmark_registry`, is that correct?
```
syntaxgym-2020
syntaxgym-2020/Licensing
syntaxgym-2020/Licensing/npi_orc_any
syntaxgym-2020/Licensing/npi_orc_ever
syntaxgym-2020/Licensing/npi_src_any
syntaxgym-2020/Licensing/npi_src_ever
syntaxgym-2020/Licensing/reflexive_orc_fem
syntaxgym-2020/Licensing/reflexive_orc_masc
syntaxgym-2020/Licensing/reflexive_prep_fem
syntaxgym-2020/Licensing/reflexive_prep_masc
syntaxgym-2020/Licensing/reflexive_src_fem
syntaxgym-2020/Licensing/reflexive_src_masc
syntaxgym-2020/Long-Distance Dependencies
syntaxgym-2020/Long-Distance Dependencies/cleft
syntaxgym-2020/Long-Distance Dependencies/cleft_modifier
syntaxgym-2020/Long-Distance Dependencies/fgd_hierarchy
syntaxgym-2020/Long-Distance Dependencies/fgd_object
syntaxgym-2020/Long-Distance Dependencies/fgd_pp
syntaxgym-2020/Long-Distance Dependencies/fgd_subject
syntaxgym-2020/Agreement
syntaxgym-2020/Agreement/number_orc
syntaxgym-2020/Agreement/number_prep
syntaxgym-2020/Agreement/number_src
syntaxgym-2020/Garden-Path Effects
syntaxgym-2020/Garden-Path Effects/mvrr
syntaxgym-2020/Garden-Path Effects/mvrr_mod
syntaxgym-2020/Garden-Path Effects/npz_ambig
syntaxgym-2020/Garden-Path Effects/npz_ambig_mod
syntaxgym-2020/Garden-Path Effects/npz_obj
syntaxgym-2020/Garden-Path Effects/npz_obj_mod
syntaxgym-2020/Gross Syntactic State
syntaxgym-2020/Gross Syntactic State/subordination
syntaxgym-2020/Gross Syntactic State/subordination_orc-orc
syntaxgym-2020/Gross Syntactic State/subordination_pp-pp
syntaxgym-2020/Gross Syntactic State/subordination_src-src
syntaxgym-2020/Center Embedding
syntaxgym-2020/Center Embedding/center_embed
syntaxgym-2020/Center Embedding/center_embed_mod
```

jimn2 commented:
@mschrimpf and @hans I just pushed a version that makes the benchmark registry look correct (I removed the top and mid-layer tiers in the hierarchy) but I think there is still a problem with it.  When I try to actually use one of the benchmarks in the new registry (i.e. number_src) I get the following error:

 `File "/Users/jimslaptop/Documents/GitHub/language/brainscore_language/plugin_management/import_plugin.py", line 43, in locate_plugin
    assert plugin_registrations_count > 0, f"No registrations found for {self.identifier}"
AssertionError: No registrations found for number_src`

I think this is related to line 172 in init.py: 

`benchmark_registry[f"{suite_name}"] = lambda: SyntaxGymSingleTSE(suite_name)` 

because suite_name is just the name of the benchmark & test_suite but the expected argument for SyntaxGymSingleTSE() is a path to the test_suite.

mschrimpf commented on brainscore_language/benchmarks/syntaxgym/__init__.py:
```diff
@@ -0,0 +1,174 @@
+      test_suite_dict = json.load(json_file)
+      suite_paths = [test_suite_dict['number_src']]
+    return SyntaxGymTSE(suite_paths)
+    # with open(Path(__file__).parent / 'test_suites.json') as json_file:
+    #     test_suite_dict = json.load(json_file)
+    #     suite_paths = list(test_suite_dict.values())
+    # return SyntaxGymTSE(suite_paths)
```
dead code

mschrimpf commented on brainscore_language/benchmarks/syntaxgym/__init__.py:
```diff
@@ -0,0 +1,174 @@
+
+for suite_name in test_suite_dict.keys():
+    # Create a benchmark with a single suite.
+    benchmark_registry[f"{suite_name}"] = lambda: SyntaxGymSingleTSE(suite_name)
```
double-check, but I think you need to be careful with the parameter binding in the lambas here: I believe ever `suite_name` benchmark will use `SyntaxGymSingleTSE` instantiated with only the _last_ `suite_name`. One work-around is to do `lambda suite_name=suite_name: SyntaxGymSingleTSE(suite_name)`

mschrimpf commented on brainscore_language/benchmarks/syntaxgym/__init__.py:
```diff
@@ -0,0 +1,174 @@
+
+for suite_name in test_suite_dict.keys():
+    # Create a benchmark with a single suite.
+    benchmark_registry[f"{suite_name}"] = lambda: SyntaxGymSingleTSE(suite_name)
```
```suggestion
    benchmark_registry[suite_name] = lambda: SyntaxGymSingleTSE(suite_name)
```
(unrelated to above comment)

jimn2 replied:
I think my last commit addresses all benchmark registry concerns.

jimn2 commented on brainscore_language/benchmarks/syntaxgym/__init__.py:
```diff
@@ -0,0 +1,44 @@
+with open(Path(__file__).parent / 'test_suites.json') as json_file:
+    test_suite_dict = json.load(json_file)
+
+benchmark_registry["syntaxgym-2020"] = SyntaxGym2020
```
Done.
